File: core/src/com/oni/donjon/actor/SaveWindow.java
package com.oni.donjon.actor;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.files.FileHandle;
import com.badlogic.gdx.scenes.scene2d.InputEvent;
import com.badlogic.gdx.scenes.scene2d.InputListener;
import com.badlogic.gdx.scenes.scene2d.ui.*;
import com.badlogic.gdx.utils.Array;
import com.badlogic.gdx.utils.Json;
import com.oni.donjon.Resources;
import com.oni.donjon.data.GameData;
import de.tomgrill.gdxdialogs.core.GDXDialogs;
import de.tomgrill.gdxdialogs.core.GDXDialogsSystem;
import de.tomgrill.gdxdialogs.core.dialogs.GDXTextPrompt;
import de.tomgrill.gdxdialogs.core.listener.TextPromptListener;
import lombok.Getter;
import lombok.Setter;

/**
 * @author Daniel Chesters (on 06/06/14).
 */
public class SaveWindow {
    @Getter
    private Window window;

    @Setter
    private MenuGameWindow menuGameWindow;

    private List<String> saveList;
    private TextButton saveButton;
    private TextButton cancelButton;
    private TextButton newSaveButton;
    public SaveWindow(String title, Skin skin) {
        this.window = new Window(title, skin);

        createSaveList(skin);
        createSaveButton(skin);
        createCancelButton(skin);
        createNewSaveButton(skin);
        createWindow(skin);
    }

    private void createWindow(Skin skin) {
        ScrollPane saveScrollPane = new ScrollPane(saveList, skin);
        saveScrollPane.setFlickScroll(false);
        window.defaults().spaceBottom(10);
        window.row();
        window.add(saveScrollPane).fill().colspan(3).maxHeight(150);
        window.row().fill().expandX();
        window.add(saveButton).colspan(1);
        window.add(newSaveButton).colspan(1);
        window.add(cancelButton).colspan(1);
        window.setModal(true);
        window.setVisible(false);
    }

    private void createSaveList(Skin skin) {
        saveList = new List<>(skin);
        saveList.getSelection().setRequired(false);
        saveList.getSelection().setMultiple(false);
    }

    private void createNewSaveButton(Skin skin) {
        newSaveButton = new TextButton(Resources.BUNDLE.get("window.save.new"), skin);
        newSaveButton.pack();
        newSaveButton.addListener(new InputListener() {
            @Override public boolean touchDown(InputEvent event, float x, float y, int pointer,
                int button) {
                GDXDialogs dialogs = GDXDialogsSystem.install();

                GDXTextPrompt textPrompt = dialogs.newDialog(GDXTextPrompt.class);

                textPrompt.setTitle(Resources.BUNDLE.get("window.save.new.input"));
                textPrompt.setMessage(Resources.BUNDLE.get("window.save.new.input"));

                textPrompt.setCancelButtonLabel(Resources.BUNDLE.get("window.save.cancel"));
                textPrompt.setConfirmButtonLabel(Resources.BUNDLE.get("window.save.ok"));

                textPrompt.setTextPromptListener(new TextPromptListener() {

                    @Override
                    public void confirm(String text) {
                        Json json = new Json();
                        FileHandle file =
                            Gdx.files.external(String.format(".config/donjon/save/%s.json", text));
                        String save = json.prettyPrint(GameData.INSTANCE);
                        file.writeString(save, false);
                        SaveWindow.this.window.setVisible(false);
                        menuGameWindow.setVisible(true);
                    }

                    @Override
                    public void cancel() {
                        SaveWindow.this.window.setVisible(false);
                        menuGameWindow.setVisible(true);
                    }
                });

                textPrompt.build().show();
                return true;
            }
        });
    }

    private void createCancelButton(Skin skin) {
        cancelButton = new TextButton(Resources.BUNDLE.get("window.save.cancel"), skin);
        cancelButton.pack();
        cancelButton.addListener(new InputListener() {
            @Override public boolean touchDown(InputEvent event, float x, float y, int pointer,
                int button) {
                SaveWindow.this.window.setVisible(false);
                menuGameWindow.setVisible(true);
                return true;
            }
        });
    }

    private void createSaveButton(Skin skin) {
        saveButton = new TextButton(Resources.BUNDLE.get("window.save.ok"), skin);
        saveButton.pack();
        saveButton.addListener(new InputListener() {
            @Override public boolean touchDown(InputEvent event, float x, float y, int pointer,
                int button) {
                String saveName = saveList.getSelected();
                Json json = new Json();
                FileHandle file =
                    Gdx.files.external(String.format(".config/donjon/save/%s", saveName));
                String save = json.prettyPrint(GameData.INSTANCE.toGameSave());
                file.writeString(save, false);
                SaveWindow.this.window.setVisible(false);
                menuGameWindow.setVisible(true);
                return true;
            }
        });
    }

    public void show() {
        FileHandle saveFolder = Gdx.files.external(".config/donjon/save");
        Array<String> saveArray = new Array<>();
        for (FileHandle file : saveFolder.list((dir, name) -> name.endsWith(".json"))) {
            saveArray.add(file.name());
        }
        saveList.setItems(saveArray);
        window.pack();
        window.setPosition(Gdx.graphics.getWidth() / 2f - window.getWidth() / 2f,
            Gdx.graphics.getHeight() / 2f - window.getHeight() / 2f);
        window.setVisible(true);
    }
}

File: core/src/main/kotlin/com/oni/donjon/actor/SaveWindow.kt
package com.oni.donjon.actor

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.scenes.scene2d.InputEvent
import com.badlogic.gdx.scenes.scene2d.InputListener
import com.badlogic.gdx.scenes.scene2d.ui.List
import com.badlogic.gdx.scenes.scene2d.ui.ScrollPane
import com.badlogic.gdx.scenes.scene2d.ui.Skin
import com.badlogic.gdx.scenes.scene2d.ui.TextButton
import com.badlogic.gdx.scenes.scene2d.ui.Window
import com.badlogic.gdx.utils.Array
import com.badlogic.gdx.utils.I18NBundle
import com.badlogic.gdx.utils.Json
import com.oni.donjon.DonjonGame
import com.oni.donjon.Resources
import com.oni.donjon.data.GameData
import de.tomgrill.gdxdialogs.core.GDXDialogsSystem
import de.tomgrill.gdxdialogs.core.dialogs.GDXTextPrompt
import de.tomgrill.gdxdialogs.core.listener.TextPromptListener
import ktx.scene2d.scene2d
import ktx.scene2d.window

/**
 * @author Daniel Chesters (on 06/06/14).
 */
class SaveWindow(title: String, private val skin: Skin, game: DonjonGame) {
    lateinit var menuGameWindow: MenuGameWindow
    private val bundle: I18NBundle = game.context.inject()

    val saveList = createSaveList()
    private val saveButton = createSaveButton()
    private val cancelButton = createCancelButton()
    private val newSaveButton = createNewSaveButton()

    val window: Window = scene2d.window(title = title, skin = skin) {
        val saveScrollPane = ScrollPane(saveList, skin)
        saveScrollPane.setFlickScroll(false)
        defaults().spaceBottom(SAVE_WINDOW_SPACE_BOTTOM)
        row()
        add(saveScrollPane).fill().colspan(SAVE_WINDOW_COLSPAN).maxHeight(SAVE_WINDOW_MAX_HEIGHT)
        row().fill().expandX()
        add(saveButton).colspan(1)
        add(newSaveButton).colspan(1)
        add(cancelButton).colspan(1)
        isModal = true
        isVisible = false
    }

    private fun createSaveList(): List<String> {
        val saveList = List<String>(skin)
        saveList.selection.required = false
        saveList.selection.multiple = false
        return saveList
    }

    private fun createNewSaveButton(): TextButton {
        val newSaveButton = TextButton(bundle["window.save.new"], skin)
        newSaveButton.pack()
        newSaveButton.addListener(object : InputListener() {
            override fun touchDown(
                event: InputEvent?,
                x: Float,
                y: Float,
                pointer: Int,
                button: Int
            ): Boolean {
                val dialogs = GDXDialogsSystem.install()

                val textPrompt = dialogs.newDialog(GDXTextPrompt::class.java)

                textPrompt.setTitle(bundle["window.save.new.input"])
                textPrompt.setMessage(bundle["window.save.new.input"])

                textPrompt.setCancelButtonLabel(bundle["window.save.cancel"])
                textPrompt.setConfirmButtonLabel(bundle["window.save.ok"])

                textPrompt.setTextPromptListener(object : TextPromptListener {

                    override fun confirm(text: String) {
                        val json = Json()
                        val file = Gdx.files.external(".config/donjon/save/$text.json")
                        val save = json.prettyPrint(GameData)
                        file.writeString(save, false)
                        this@SaveWindow.window.isVisible = false
                        menuGameWindow.setVisible(true)
                    }

                    override fun cancel() {
                        this@SaveWindow.window.isVisible = false
                        menuGameWindow.setVisible(true)
                    }
                })

                textPrompt.build().show()
                return true
            }
        })
        return newSaveButton
    }

    private fun createCancelButton(): TextButton {
        val cancelButton = TextButton(bundle["window.save.cancel"], skin)
        cancelButton.pack()
        cancelButton.addListener(object : InputListener() {
            override fun touchDown(
                event: InputEvent?,
                x: Float,
                y: Float,
                pointer: Int,
                button: Int
            ): Boolean {
                this@SaveWindow.window.isVisible = false
                menuGameWindow.setVisible(true)
                return true
            }
        })

        return cancelButton
    }

    private fun createSaveButton(): TextButton {
        val saveButton = TextButton(bundle["window.save.ok"], skin)
        saveButton.pack()
        saveButton.addListener(object : InputListener() {
            override fun touchDown(
                event: InputEvent?,
                x: Float,
                y: Float,
                pointer: Int,
                button: Int
            ): Boolean {
                val saveName = saveList.selected
                val json = Json()
                val file = Gdx.files.external(".config/donjon/save/$saveName")
                val save = json.prettyPrint(GameData.toGameSave())
                file.writeString(save, false)
                this@SaveWindow.window.isVisible = false
                menuGameWindow.setVisible(true)
                return true
            }
        })
        return saveButton
    }

    fun show() {
        val saveFolder = Gdx.files.external(".config/donjon/save")
        val saveArray = Array<String>()
        for (file in saveFolder.list { _, name -> name.endsWith(".json") }) {
            saveArray.add(file.name())
        }
        saveList.setItems(saveArray)
        window.pack()
        window.setPosition(
            Gdx.graphics.width / 2f - window.width / 2f,
            Gdx.graphics.height / 2f - window.height / 2f
        )
        window.isVisible = true
    }

    companion object {
        const val SAVE_WINDOW_SPACE_BOTTOM = 10f
        const val SAVE_WINDOW_COLSPAN = 3
        const val SAVE_WINDOW_MAX_HEIGHT = 150f
    }
}
